scripts/image_drift.sh

```diff
@@ -0,0 +1,161 @@
+#!/bin/bash
+set -euo pipefail
+
+# --- Input Validation and Configuration ---
+if [ "$#" -ne 1 ]; then
+  echo "Usage: $0 <UPSTREAM_CHART_PATH>"
+  echo "Example (All Charts): $0 ./helmfile.d/upstream"
+  echo "Example (Single Chart): $0 ./helmfile.d/upstream/grafana/grafana"
+  exit 1
+fi
+
+INPUT_PATH="$1"
+IMAGES_YAML_PATH="./helmfile.d/lists/images.yaml"
+IMAGE_MAP_FILE="./tests/unit/general/resources/images-parametric-tests.json"
+
+if [ ! -f "$IMAGE_MAP_FILE" ]; then
+  echo "ERROR: Mapping file '$IMAGE_MAP_FILE' not found. Please ensure it is in the same directory."
+  exit 1
+fi
+
+update_images_yaml_tag() {
+  local key="$1"
+  local current_ref="$2"
+  local new_tag="$3"
+
+  local repo="${current_ref%:*}"
+  local updated="${repo}:${new_tag}"
+  # shellcheck disable=SC2016
+  yq -i ".images.$key = \"$updated\"" "$IMAGES_YAML_PATH"
+  echo "UPDATED: images.yaml -> .images.$key = $updated"
+}
+
+# --- Determine Scope (Single Chart or All Charts) and Calculate Base Directory ---
+LAST_PATH_COMPONENT=$(basename "$INPUT_PATH")
+CHART_FILTER=""
+UPSTREAM_BASE_DIR=""
+
+if [ -f "$INPUT_PATH/values.yaml" ]; then
+  # Single Chart Mode: Calculate UPSTREAM_BASE_DIR
+  CHECK_SCOPE="Single Chart: $LAST_PATH_COMPONENT"
+  CHART_FILTER="$LAST_PATH_COMPONENT"
+
+  for MAP_OBJECT in $(jq -c '.parameters[]' "$IMAGE_MAP_FILE"); do
```

Comment:
Consider using `readarray` to guard against accidental word splitting in the middle of lines.